### ISSUE TITLE
Support for Other Types of Tensors in tf.data (For example RaggedTensors / Sparse Tensors)

### ISSUE THREAD

**Issue by captain-pool** [tensorflow/tensorflow#27653]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Currently tf.data only supports Regular Tensors.
**Will this change the current api? How?**
This will not change the API on a large scale. This will just allow tf.data to return RaggedTensor or SparseTensors depending on the Users Need. This will save a lot of memory when it comes to reading Variable Length Sequences.
**Who will benefit with this feature?**
Anyone Who is using tensorflow.
**Any Other info.**
I did some digging in the repository, while trying to manually trace function calls and see which functions and classes are needed to be updated.
1. A similar Overload of GetNextInternal is needed in dataset.h for every new type of vector of Tensor to be Supported.(Sparse and Ragged) https://github.com/tensorflow/tensorflow/blob/222fea0388a9ca5ef5e736156e6604fbe51f07b0/tensorflow/core/framework/dataset.h#L840-L842
2. A new similar overload of GetNext() with new type of vector of the type of Tensor to be supported (Sparse and Ragged) https://github.com/tensorflow/tensorflow/blob/222fea0388a9ca5ef5e736156e6604fbe51f07b0/tensorflow/core/framework/dataset.h#L720-L721
3. Implement this GetNextInternal definition in the data kernel ops. Similar to these
https://github.com/tensorflow/tensorflow/blob/222fea0388a9ca5ef5e736156e6604fbe51f07b0/tensorflow/core/kernels/data/interleave_dataset_op.cc#L162-L164
4. Create modifications of Python Interfaces for these kernels, here https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/ops/dataset_ops.py

**Comment by captain-pool** [comment]
I'ld like to work on this issue / Collaborate.
@jsimsa

**Comment by jsimsa** [comment]
@captain-pool your statement that tf.data only supports regular tensors is not true.

It already supports `tf.SparseTensor`s and there is an internal PR in progress that will add support for `tf.RaggedTensor`s.

Note that the support exists at Python level and is provided through the Structure module (for instance see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/util/structure.py#L501).

The idea is that non-dense tensors are boxed as regular tensors when they are passed into op kernels because in C++ there are no counterparts to `tf.SparseTensor` or `tf.RaggedTensor`. The added benefit of doing that is that we do not have to write N kernels for each tf.data op, where N would be the number of types to support.

In other words, this feature request is already partially addressed and will be completely address once the RaggedTensor support is completed. For the time being, I am assigning the to @edloper who is leading the RaggedTensor effort.

**Comment by captain-pool** [comment]
> @captain-pool your statement that tf.data only supports regular tensors is not true.
> 
> It already supports `tf.SparseTensor`s and there is an internal PR in progress that will add support for `tf.RaggedTensor`s.
> 
> Note that the support exists at Python level and is provided through the Structure module (for instance see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/util/structure.py#L501).
> 
> The idea is that non-dense tensors are boxed as regular tensors when they are passed into op kernels because in C++ there are no counterparts to `tf.SparseTensor` or `tf.RaggedTensor`. The added benefit of doing that is that we do not have to write N kernels for each tf.data op, where N would be the number of types to support.
> 
> In other words, this feature request is already partially addressed and will be completely address once the RaggedTensor support is completed. For the time being, I am assigning the to @edloper who is leading the RaggedTensor effort.

Thanks for Clearing out the Doubts. @jsimsa will definitely look into it.

**Comment by edloper** [comment]
tf.data support for RaggedTensors was added by 5fe90dc.

**Comment by PistaSaki** [comment]
Thank you for your work!
However I think, there is a bug somewhere. I tried the following code in tensorflow 2.0.0-beta0:
```python
import tensorflow as tf
from tensorflow.data import Dataset

data = tf.ragged.constant([
        [[1, 1],
         [2, 2]],
        
        [[3, 3]],
         
        [],
        
        [[4, 4],
         [5, 5]]
    ], dtype=tf.float32, ragged_rank=1)

ds = Dataset.from_tensor_slices(data)
ds = ds.batch(2)

for i, batch in enumerate(ds):
    print("batch", i)
    for x in batch:
        print(x.numpy())
```
The output is:
```
batch 0
[[1. 1.]
 [2. 2.]
 [3. 3.]]
[]
batch 1
[]
[[4. 4.]
 [5. 5.]]
```
the correct output should be:
```
batch 0
[[1. 1.]
 [2. 2.]]
[[3. 3.]]
batch 1
[]
[[4. 4.]
 [5. 5.]]
```

**Comment by edloper** [comment]
Thanks for pointing this out -- I agree that something looks broken there.
We'll look into it and get back to you.

Thanks,
-Edward

**Comment by hamatake** [comment]
Is this fixed now?

**Comment by edloper** [comment]
Yes this was fixed by 84cf4d436eb5e41b0a7b24a3a079fd20c1603594.

**Comment by PistaSaki** [comment]
Thank you, it seems to work!
I have a related problem, but maybe I should open it as another issue:
I have a dataset from generator, containing uneven tensors. After batching I would expect it to return dataset with RaggedTensors. At present it fails.
Here is a minimal example:
```python
import numpy as np
import tensorflow as tf
from tensorflow.data import Dataset

def get_iterator():
    return map(lambda x: np.ones([x, 1]), range(4))
    
ds = Dataset.from_generator(get_iterator, output_types= tf.float32,
                            output_shapes=[None, 1])

for x in ds:
    print(x)    

print("Batched:")

for x in ds.batch(2):
    print(x) 
```

**Comment by VitaliKaiser** [comment]
@PistaSaki 

If this is still open issue (and if somebody else like myself is looking this up):
from_generator creates regular tf.Tensors.
But you could explicitly cast them to tf.RaggedTensor:

```python
import numpy as np
import tensorflow as tf
from tensorflow.data import Dataset

def get_iterator():
    return map(lambda x: np.ones([x, 1]), range(4))
    
ds = Dataset.from_generator(get_iterator, output_types= tf.float32,
                            output_shapes=[None, 1])

ds = ds.map(tf.RaggedTensor.from_tensor)

for x in ds:
    print(x)    

print("Batched:")

for x in ds.batch(2):
    print(x) 
```

**Comment by edloper** [comment]
You can use [tf.data.experimental.dense_to_ragged_batch](https://www.tensorflow.org/api_docs/python/tf/data/experimental/dense_to_ragged_batch) to batch variable-sized tensors into a ragged tensor.

**Comment by Ceceu** [comment]
@PistaSaki, @edloper,
Is there an approach that returns a `tf.RaggedTensor` using `tf.data.Dataset.from_generator` instead of a `tf.Tensor`?

**Comment by edloper** [comment]
@PistaSaki, @VitaliKaiser If you have dense data and want to batch it to become ragged, you can use [`tf.data.experimental.dense_to_ragged_batch`](https://www.tensorflow.org/api_docs/python/tf/data/experimental/dense_to_ragged_batch?version=nightly).

@Ceceu There is a pull request (#37400) that will update `Dataset.from_generator` to work with `RaggedTensor` (or other composite tensors).  But I think it's been stalled for a little while on making sure it doesn't break a test.
